windows-agent/internal/proservices/landscape/landscape_test.go

```diff
@@ -540,6 +538,157 @@ func TestAutoReconnection(t *testing.T) {
+		c.mu.Unlock()
+
+		c.triggerNotifications()
+	}
+
+	changeCertificate := func(ctx context.Context, s *landscape.Service, c *mockConfig) {
+		// We change the path to an equivalent one, so that the reconnect is triggered and the connection still succeeds
+		const sep = filepath.Separator
+		from := fmt.Sprintf("%c", sep)       // from: /
+		to := fmt.Sprintf("%c.%c", sep, sep) // to:   /./
+
+		c.mu.Lock()
+		c.landscapeClientConfig = strings.Replace(c.landscapeClientConfig, from, to, 1)
+		c.mu.Unlock()
+
+		c.triggerNotifications()
+	}
+
+	changeIrrelevant := func(ctx context.Context, s *landscape.Service, c *mockConfig) {
+		c.mu.Lock()
+		c.landscapeClientConfig = c.landscapeClientConfig + "\n[exta]\ninfo=this section does not matter"
+		c.mu.Unlock()
+
+		c.triggerNotifications()
+	}
+
+	testCases := map[string]struct {
+		useCertificate bool
+		trigger        func(context.Context, *landscape.Service, *mockConfig)
+
+		wantNoReconnect       bool
+		wantImmediateRconnect bool
+	}{
+		"Reconnect when explicitly requesting a reconnection": {trigger: requestReconnect, wantImmediateRconnect: true},
+		"Reconnect when changing the URL":                     {trigger: changeAddress},
+		"Reconnect when changing the certificate path":        {trigger: changeCertificate, useCertificate: true},
+		"Don't reconnect when changing irrelevant config":     {trigger: changeIrrelevant, wantNoReconnect: true},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			if wsl.MockAvailable() {
+				ctx = wsl.WithMock(ctx, wslmock.New())
+			}
+
+			var certPath string
+			lcapeConfig := defaultLandscapeConfig
+			if tc.useCertificate {
+				certPath = t.TempDir()
+				testutils.GenerateTempCertificate(t, certPath)
+				lcapeConfig = fmt.Sprintf("%s\nssl_public_key = {{ .CertPath }}/cert.pem", defaultLandscapeConfig)
+			}
+
+			lis, server, _ := setUpLandscapeMock(t, ctx, "localhost:", certPath)
+
+			conf := &mockConfig{
+				proToken:              "TOKEN",
+				landscapeClientConfig: executeLandscapeConfigTemplate(t, lcapeConfig, certPath, lis.Addr()),
+			}
+
+			//nolint:errcheck // We don't care about these errors
+			go server.Serve(lis)
+			defer server.Stop()
+
+			db, err := database.New(ctx, t.TempDir(), conf)
+			require.NoError(t, err, "Setup: database New should not return an error")
+
+			service, err := landscape.New(ctx, conf, db)
+			require.NoError(t, err, "Setup: New should not return an error")
+
+			err = service.Connect()
+			require.NoError(t, err, "Setup: Connect should not return an error")
+
+			require.Eventually(t, func() bool {
+				return service.Connected()
+			}, 5*time.Second, 500*time.Millisecond, "Client should have reconnected after restarting the server")
+
+			ch := make(chan struct{})
+			go func() {
+				tc.trigger(ctx, service, conf)
+				close(ch)
+			}()
+
+			fmt.Println("AAAAAAAAA")
```

Comment:
left-over after some printf-debugging section?

Reply:
Hahaha you can see my desperation there